Bottom: 7ea673b6952289368b8054eff5fc190c412746f8
Top:    e7ed8f0b714cef70e4b910f7812a8476457e15b1
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-09-05 13:53:56 +0100

hid/gtk: Use gtk_widget_get_allocation() not gdk_window_get_geometry()

This avoids an unnecessary round-trip the the X server on X11 platforms.


---

diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index 21a22e8..13c45b0 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -1028,7 +1028,7 @@ ghid_pinout_preview_expose (GtkWidget *widget,
   GdkGLContext* pGlContext = gtk_widget_get_gl_context (widget);
   GdkGLDrawable* pGlDrawable = gtk_widget_get_gl_drawable (widget);
   GhidPinoutPreview *pinout = GHID_PINOUT_PREVIEW (widget);
-  int da_w, da_h;
+  GtkAllocation allocation;
   view_data save_view;
   int save_width, save_height;
   double xz, yz;
@@ -1039,18 +1039,18 @@ ghid_pinout_preview_expose (GtkWidget *widget,
 
   /* Setup zoom factor for drawing routines */
 
-  gdk_window_get_geometry (widget->window, 0, 0, &da_w, &da_h, 0);
-  xz = (double) pinout->x_max / da_w;
-  yz = (double) pinout->y_max / da_h;
+  gtk_widget_get_allocation (widget, &allocation);
+  xz = (double) pinout->x_max / allocation.width;
+  yz = (double) pinout->y_max / allocation.height;
   if (xz > yz)
     gport->view.coord_per_px = xz;
   else
     gport->view.coord_per_px = yz;
 
-  gport->width = da_w;
-  gport->height = da_h;
-  gport->view.width = da_w * gport->view.coord_per_px;
-  gport->view.height = da_h * gport->view.coord_per_px;
+  gport->width = allocation.width;
+  gport->height = allocation.height;
+  gport->view.width = allocation.width * gport->view.coord_per_px;
+  gport->view.height = allocation.height * gport->view.coord_per_px;
   gport->view.x0 = (pinout->x_max - gport->view.width) / 2;
   gport->view.y0 = (pinout->y_max - gport->view.height) / 2;
